src/System.Management.Automation/engine/Interop/Windows/SetInformationJobObject.cs

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#nullable enable
+
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static unsafe partial class Windows
+    {
+        internal const int JobObjectAssociateCompletionPortInformation = 7;
+        internal const int JOB_OBJECT_MSG_ACTIVE_PROCESS_ZERO = 4;
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct JOBOBJECT_ASSOCIATE_COMPLETION_PORT
+        {
+            public nint CompletionKey;
+            public nint CompletionPort;
+        }
+
+        [LibraryImport("kernel32.dll", SetLastError = true)]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        internal static partial bool SetInformationJobObject(
+            nint hJob,
+            int JobObjectInformationClass,
+            nint lpJobObjectInformation,
+            int cbJobObjectInformationLength);
+
+        internal static bool SetInformationJobObject(
+            nint jobHandle,
+            ref JOBOBJECT_ASSOCIATE_COMPLETION_PORT completionPort)
+        {
+            unsafe
```

Comment:
See example FindFirstFileW in our interop code how avoid unsafe code.

Reply:
I personally find it easier to reason with a small unsafe block then having multiple definition for a PInvoke call just to expose the various `ref` overloads. In this case having this overload with the unsafe block means that

+ the caller doesn't have to worry about also passing in the `JobObjectInformationClass` and `cbJobObjectInformationLength`
+ we don't need to duplicate the `LibraryImport` definition

While the first point isn't inherent to this using unsafe it does mean we can avoid the latter through this more specific overload.

Reply:
LibraryImport automatically involves a source generator to create specific marshaling code. If we allow the SG to process SafeHandle why do we manually process the struct? The SG creates well-tested code for us. Main benefit of the SG is that we automatically get all innovations from .Net developers. With a manual code, we lose that. The point is that using the SG is preferable, as we have already done in other places.

The struct is only for the pinvoke but it is created in main code. Since we already have the helper method I'd expect the struct creation in the method.

The pinvoke method is only used in the helper method so it should be private.

Reply:
Fair enough, I'll just add that one definition considering we only use it for that right now anyway.